Disable additional network components when most are disabled, in all platforms in OvmfPkg and ArmVirtPkg

mikebeaton opened the pull request:
# Description

This PR resolves the other issue identified in https://github.com/tianocore/edk2/pull/6087 - namely that various other platforms were only not showing the same problem, of failing to build when -D NETWORK_ENABLE=0 is specified, because they were incorrectly still including some NetworkPkg components when most were disabled by -D NETWORK_ENABLE=0 (which masks the issue in https://github.com/tianocore/edk2/pull/6087). To identify the changes to make, it was necessary to identify which NetworkPkg components were being unnecessarily included, and then to add conditional include statements for the affected components (all of which were already conditionally included in the Intel OvmfPkg platforms, after changes introduced in d933ec115bdf9be1d8dfe6a818414a14973cc0d3 and 7f17a155640a2a9e1f7b0f3522628ee2c6f62624).

**EDIT**: I have now split the original, single cross-package commit into two. On the request/suggestion below, I have now also added ~two~ further commits which implement the change of using the `OvmfPkg/Include/*/Shell*.inc` files (which already had this logic and were used in some platforms), instead of just copying the logic from them. ~These changes may well be worthwhile, though (as I somewhat suspected) they were slightly less straightforward to implement than might have been expected. Additional comments on these extra commits below.~ (After a pointer from @kraxel as to how the shell+secure boot+CI thing worked in OvmfPkg, things became more straightforward. :-))

- [ ] Breaking change? NO
- [ ] Impacts security? NO
- [ ] Includes tests? NO

## How This Was Tested

Identify  all NetworkPkg components still being included when compiled with -D NETWORK_ENABLE=0, in platforms which were already using NetworkPkg includes intended to allow all NetworkPkg components to be disabled via -D NETWORK_ENABLE=0.

After the changes to avoid these residual NetworkPkg includes, test that all the platforms affected would _not_ build with -D NETWORK_ENABLE=0 (as already happened in the Intel Ovmf platforms, on which similar changes to avoid residual NetworkPkg includes had been made by d933ec115bdf9be1d8dfe6a818414a14973cc0d3 followed by 7f17a155640a2a9e1f7b0f3522628ee2c6f62624). This confirms that no NetworkPkg components are being included any more. Confirm that these platforms still build the same as before with -D NETWORK_ENABLE=1 (default), and still pass CI.

## Integration Instructions

The paired change https://github.com/tianocore/edk2/pull/6087 (which can be applied before or after this) resolves the remaining issue of access to undefined Pcds and allows the platforms changed here to build again with -D NETWORK_ENABLE=0.

mikebeaton commented:
If I may, cc @SaloniKasbekar @weltling @mdkinney @kraxel - as commented above, one of two commits here is as you already approved/showed interest in approving. Thank you. Does the other commit look sensible to include?

mikebeaton commented:
Having now identified that d933ec115bdf9be1d8dfe6a818414a14973cc0d3 is also relevant (i.e. along with 7f17a155640a2a9e1f7b0f3522628ee2c6f62624 was part of removing all references to residual NetworkPkg components in the Intel OvmfPkg platforms), I wanted to update the commit messages to also mention this. I would still like to do so, but I have perhaps done enough force pushes for now, so will just mention here that that update is pending.

mikebeaton commented:
I have realised it makes more sense for the change here to be a separate PR, so have put this back to just one commit and have re-opened https://github.com/tianocore/edk2/pull/6087 (with the minor update to add LoongArchVirtQemu, which is also affected).

kraxel commented:
How about switching the builds over to use the `OvmfPkg/Include/*/Shell*.inc` include files (which already have this logic)?

mikebeaton commented:
@kraxel - The existing includes are not applicable everywhere. These platforms all have slight variants of what's included, so switching to those includes in the obvious places wouldn't be like-for-like.

https://github.com/tianocore/edk2/blob/master/OvmfPkg/Include/Dsc/NetworkComponents.dsc.inc#L18-L34 has additional library and entry point settings for Ip4Dxe.inf and Ip6Dxe.inf. These settings aren't currently applied at any of the other places where I might otherwise want to include the top lines from this file (search for `UefiPxeBcDxe` in Files Changed of this PR).

https://github.com/tianocore/edk2/blob/master/OvmfPkg/Include/Fdf/ShellDxe.fdf.inc matches what is included in some but not all of the places where I might want to use it (search for `ShellPkg/DynamicCommand/TftpDynamicCommand/TftpDynamicCommand.inf` within `.fdf` files in Files Changed). Might also require `BUILD_SHELL` or `SECURE_BOOT_ENABLE` defines to be present, if they are not already in the other packages?

https://github.com/tianocore/edk2/blob/master/OvmfPkg/Include/Dsc/ShellComponents.dsc.inc similar to the above: some of the places where I might approximately want to use it only include `TftpDynamicCommand` (not both `TftpDynamicCommand` and `HttpDynamicCommand`), so not like-for-like. Also, this file is the full Shell components list, so again even where it is clearly more or less the same thing as what it would (approximately) replace, the exact components imported (and possibly the defines present, to trigger the correct behaviour of the include file?) are not exactly the same between the packages.

Obviously with some updates and care (and preferably running versions of all the platforms being changed, to be sure it's all working; I tend to use OvmfPkg Intel variants, and have dabbled with running ArmVirtPkg, so far!), this could be done, but not sure if this is enough to render all this possibly more suitable for future work?

kraxel commented:
I don't think the `ShellPkg/DynamicCommand/*`  differences are intentional, they probably sneaked in over time as commands have been added and not all *.dsc files have been updated.

ardbiesheuvel commented on ArmVirtPkg/ArmVirtQemu.dsc:
```diff
@@ -502,16 +502,18 @@
+      <LibraryClasses>
+        NULL|OvmfPkg/Library/TlsAuthConfigLib/TlsAuthConfigLib.inf
+    }
+  !endif
```
Please be consistent with the indentation: below, there is a nested NETWORK_ENABLE/NETWORK_IP4_ENABLE conditional where the indentation is untouched, whereas here, you indent the inner conditional along with the lines it guards.

I prefer the NETWORK_ENABLE/NETWORK_IP4_ENABLE approach, because it makes the diff easier to read - please update these hunks accordingly.

mikebeaton replied:
While noting that ofc GitHub allows showing or hiding whitespace-only changes in diffs, the main point here is that I have copied the style of indentation used for the same change, when already made elsewhere in the commits I mentioned:

https://github.com/tianocore/edk2/commit/d933ec115bdf9be1d8dfe6a818414a14973cc0d3
https://github.com/tianocore/edk2/commit/7f17a155640a2a9e1f7b0f3522628ee2c6f62624

If I keep all the indentations left aligned (not additionally indented) in this PR, then the layout will be inconsistent with the existing changed files.

I thought (& would still try to argue...) that it is preferable to keep a consistent layout between files. Though can change if you still disagree. (Not sure to what extent this will become moot anyway, if I get the time to make the additional changes @kraxel requests.)

mikebeaton replied:
For now, in the split commits, I have stuck to using the indentation introduced in https://github.com/tianocore/edk2/commit/d933ec115bdf9be1d8dfe6a818414a14973cc0d3 where I have duplicated the changes made in https://github.com/tianocore/edk2/commit/d933ec115bdf9be1d8dfe6a818414a14973cc0d3 (by @ardbiesheuvel), and using the indentation from https://github.com/tianocore/edk2/commit/7f17a155640a2a9e1f7b0f3522628ee2c6f62624 where I have duplicated the changes made in https://github.com/tianocore/edk2/commit/7f17a155640a2a9e1f7b0f3522628ee2c6f62624 (by @kraxel), even though these indentation styles are different!

mikebeaton replied:
Still finishing sorting out kraxel's request. Will make the required changes here too.

mikebeaton commented:
> How about switching the builds over to use the `OvmfPkg/Include/*/Shell*.inc` include files (which already have this logic)?

@kraxel [This](https://github.com/tianocore/edk2/pull/6187) is an initial attempt to do the above correctly. I am not yet quite sure why a few CI tests are failing, if you have any pointers? If I build and run X64 GCC ArmVirtQemu by hand, from that commit, then it reaches the UEFI Shell correctly after pausing in PXE boot for a bit - I believe as expected.

mikebeaton commented:
@kraxel @ardbiesheuvel -

I have split the previous cross-package commit into three commits.

I have also implemented the suggested change of using the `OvmfPkg/Include/*/Shell*.inc` files (which already had this logic and were using in some platforms), in two further commits. As can be seen, this was not as trivial as it might perhaps have sounded - which is the main reason that it seems right to keep the initial changes (which actually stop unwanted network components from being included, without otherwise changing anything) separate from the the somewhat more invasive changes involved in sharing the include files.

kraxel commented:
> **EDIT 2**: Right, I believe the issue is that the build matrix defines SECURE_BOOT_ENABLE, for AARCH64 only (https://dev.azure.com/tianocore/edk2-ci/_build/results?buildId=149296&view=logs&j=a3cbe7a7-5fa9-5965-2615-b39fbc4e581a and expand 'Job preparation parameters'). Since the include files (more or less correctly, I'd say) don't include UEFI Shell in the FDF when SECURE_BOOT_ENABLE is TRUE, the 'Run to shell' test for AARCH64 now fails.

See https://github.com/tianocore/edk2/commit/bc982869dd3e how this works for OVMF.

mikebeaton commented on ArmVirtPkg/ArmVirtKvmTool.fdf:
```diff
@@ -173,8 +173,7 @@ READ_LOCK_STATUS   = TRUE
   INF OvmfPkg/VirtioScsiDxe/VirtioScsi.inf
   INF OvmfPkg/VirtioRngDxe/VirtioRng.inf
 
-  INF ShellPkg/Application/Shell/Shell.inf
-  INF ShellPkg/DynamicCommand/TftpDynamicCommand/TftpDynamicCommand.inf
```
@kraxel - This is probably the most 'invasive' of the changes similar to this, as it's less clear (to me) whether it was intentional that this FDF includes neither of the 'extra' two commands (`VariablePolicyDynamicCommand` and `LinuxInitrdDynamicShellCommand`; as well as not including `HttpDynamicCommand`)

It is hopefully (at worst) harmless (i.e. possibly it's a good thing, after all!) if these are included. If not then an additional define to disable these (possibly only actually used in this specific platform) could be added.

mikebeaton replied:
On further examination as to what ArmVirtKvmTool is (i.e. just a version of ArmVirtPkg to run in the lightweight kvmtool hypervisor) then it seems fairly unlikely that there is any issue synchronising the commands here with the other OvmfPkg and ArmVirtPkg platforms.

mikebeaton replied:
PS Of course it is tested to build and run fine in CI. If I grab a chance I'll check it by hand, but it seems fine.

mikebeaton commented:
> > **EDIT 2**: Right, I believe the issue is that the build matrix defines SECURE_BOOT_ENABLE, for AARCH64 only (https://dev.azure.com/tianocore/edk2-ci/_build/results?buildId=149296&view=logs&j=a3cbe7a7-5fa9-5965-2615-b39fbc4e581a and expand 'Job preparation parameters'). Since the include files (more or less correctly, I'd say) don't include UEFI Shell in the FDF when SECURE_BOOT_ENABLE is TRUE, the 'Run to shell' test for AARCH64 now fails.
> 
> See [bc982869dd3e](https://github.com/tianocore/edk2/commit/bc982869dd3e) how this works for OVMF.

Ah, right, I didn't spot this until just now and came up with my own different solution - which I think is actually reasonable. *Possibly* even worth considering using instead? (&/or of course, I can start up again, and switch to using this method in the places where I had problems.)

mikebeaton commented:
> > **EDIT 2**: Right, I believe the issue is that the build matrix defines SECURE_BOOT_ENABLE, for AARCH64 only (https://dev.azure.com/tianocore/edk2-ci/_build/results?buildId=149296&view=logs&j=a3cbe7a7-5fa9-5965-2615-b39fbc4e581a and expand 'Job preparation parameters'). Since the include files (more or less correctly, I'd say) don't include UEFI Shell in the FDF when SECURE_BOOT_ENABLE is TRUE, the 'Run to shell' test for AARCH64 now fails.
> 
> See [bc982869dd3e](https://github.com/tianocore/edk2/commit/bc982869dd3e) how this works for OVMF.

@kraxel - I think I get it. When secure boot is enabled the shell is built, even though not bundled into firmware, as much as anything to allow the above approach to work in CI? I'll look into not breaking that(!), and adding it to ArmVirtPkg as needed.

kraxel commented:
> > See [bc982869dd3e](https://github.com/tianocore/edk2/commit/bc982869dd3e) how this works for OVMF.
> 
> @kraxel - I think I get it. When secure boot is enabled the shell is built, even though not bundled into firmware, as much as anything to allow the above approach to work in CI? I'll look into not breaking that(!), and adding it to ArmVirtPkg as needed.

Exactly, the idea is to build the shell unconditionally so CI can use it, but exclude it from the firmware images if secure boot is enabled.

mikebeaton commented:
@kraxel - I've switched to using the existing approach to handling Secure Boot + UEFI Shell + CI.

@ardbiesheuvel - I've fixed the whitespace issues.

mikebeaton commented:
Is this PR waiting for further approvals?

mdkinney commented:
Need some OvmfPkg approvals

tlendacky commented:
I'm not the owner / reviewer of the OvmfPkg files that were changed, so not sure why I would be a required reviewer. But if you really need my review approval, I guess I can do that, or you can remove me from the list.

tperard reviewed (approved):
The changes seems to work fine for OvmfXen. Thanks.

vlsunil reviewed (changes requested):
RiscVVirt looks good to me.

vlsunil reviewed (approved):
RiscVVirt looks good to me.

lgao4 commented:
@ardbiesheuvel , is there the quality risk to merge them for this table tag 202411? This patch has been reviewed before soft feature feature freeze. It can be merged for 202411.

mikebeaton commented:
Updated to fix minor merge conflicts with b3b3cfab7eb52acd77558a9727196e30d05d1f2a (cc @kraxel) and with addition of NETWORK_PXE_BOOT_ENABLE (https://github.com/tianocore/edk2/commits?author=xpahos cc @xpahos).

Have also updated the commit messages of the final two commits to be slightly more confident in tone.

mikebeaton commented:
Fix for minor merge conflict with https://github.com/tianocore/edk2/commit/b8f3199595d23c29433528a5207a6aa9fb368d44 cc @kilaterlee .

Is there any chance this could be merged? :-)

mdkinney commented:
@ardbiesheuvel Can this be merged now?
